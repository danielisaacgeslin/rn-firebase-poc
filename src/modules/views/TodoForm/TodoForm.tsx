import React, { useState, memo, useCallback, useMemo } from 'react';
import { View, Text, TextInput, KeyboardAvoidingView } from 'react-native';
import { Button } from 'react-native-elements';
import { useNavigationEvents, useNavigation } from 'react-navigation-hooks';

import { ENV, STYLE } from '../../../constants';
import { TodoModel } from '../../models';
import styles from './styles';

export interface ITodoFormProps {
  createTodo: (todo: TodoModel.ITodo) => void;
}

export interface ITodoFormState {
  title: string;
  description: string;
}

const TodoForm = ({ createTodo }: ITodoFormProps) => {
  const [state, setState] = useState<ITodoFormState>({ title: '', description: '' });
  const { navigate } = useNavigation();

  const isDisabled = useMemo(() => !state.title, [state.title]);
  const onChangetitle = useCallback((title: string) => setState({ ...state, title }), [state, setState]);
  const onChangedescription = useCallback((description: string) => setState({ ...state, description }), [state, setState]);
  const onCreate = useCallback(() => createTodo({ title: state.title, description: state.description, status: TodoModel.Status.PENDING }), [createTodo, state]);
  const goToList = useCallback(() => navigate('TodoList'), [state, setState]);

  useNavigationEvents(event => {
    if (event.type === 'willFocus') setState({ ...state, title: '', description: '' });
  });

  return (
    <View style={styles.container}>
      <KeyboardAvoidingView behavior="padding" style={styles.keyboardContainer} keyboardVerticalOffset={ENV.KEYBOARD_VERTICAL_OFFSET}>
        <View style={styles.formContainer}>
          <View>
            <Text style={styles.heading1}>Create a Todo</Text>
            <View style={styles.inputContainer}>
              <Text style={styles.inputLabel}>title</Text>
              <TextInput
                placeholder="Enter your title..."
                placeholderTextColor={STYLE.COLOR.OCTANARY}
                style={styles.input}
                underlineColorAndroid="rgba(0, 0, 0, 0)"
                value={state.title}
                onChangeText={onChangetitle}
                testID="title-input"
              />
            </View>
            <View style={styles.inputContainer}>
              <Text style={styles.inputLabel}>description</Text>
              <TextInput
                placeholder="Enter your description..."
                placeholderTextColor={STYLE.COLOR.OCTANARY}
                style={styles.input}
                underlineColorAndroid="rgba(0, 0, 0, 0)"
                value={state.description}
                onChangeText={onChangedescription}
                testID="description-input"
              />
            </View>
          </View>
          <View>
            <Button disabled={isDisabled} onPress={onCreate} title="Create Todo" testID="create-button" />
            <Button onPress={goToList} title="Go to List" type="clear" testID="go-to-list-button" buttonStyle={{ marginTop: 15 }} />
          </View>
        </View>
      </KeyboardAvoidingView>
    </View>
  );
};

export default memo(TodoForm);
