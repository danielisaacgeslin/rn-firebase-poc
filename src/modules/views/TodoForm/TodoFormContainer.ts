import { connect } from 'react-redux';

import { IRootState } from '../../state-mgmt/rootState';
import TodoForm from './TodoForm';
import { TodoModel } from '../../models';
import { todoState } from '../../state-mgmt/todo';

export const mapStateToProps = (state: IRootState) => ({

});

export const mapDispatchToProps = dispatch => ({
  createTodo: (todo: TodoModel.ITodo) => dispatch(todoState.actions.createStart(todo))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoForm);
