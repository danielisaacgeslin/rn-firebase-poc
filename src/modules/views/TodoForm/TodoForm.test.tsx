import React, { Children } from 'react';
import { fireEvent, render, RenderAPI } from 'react-native-testing-library';
import { createSwitchNavigator, NavigationParams, createAppContainer } from 'react-navigation';

import TodoForm, { ITodoFormProps } from './TodoForm';
import { TodoModel } from '../../models';

const createTestNavigator = (ref: React.RefObject<any>) => ({ children, params }: { children: any; params?: NavigationParams }) => {
  const Navigator = createAppContainer(
    createSwitchNavigator({
      TestScreen: { screen: () => Children.only(children), params } as any,
      TodoList: () => <div>todo list</div> // to test navigation to this screen
    })
  );
  return <Navigator ref={ref} />;
};

describe('TodoForm', () => {
  global.console.warn = () => null;
  global.console.error = () => null;
  let props: ITodoFormProps;
  let wrapper: RenderAPI;
  let TestNavigator;
  let navRef;

  beforeEach(() => {
    navRef = React.createRef();
    TestNavigator = createTestNavigator(navRef);
    props = {
      createTodo: jest.fn()
    };

    wrapper = render(
      <TestNavigator>
        <TodoForm {...props} />
      </TestNavigator>
    );
  });

  it('should render', () => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should create a todo', () => {
    fireEvent.changeText(wrapper.getByTestId('title-input'), 'title');
    fireEvent.changeText(wrapper.getByTestId('description-input'), 'description');
    fireEvent.press(wrapper.getByTestId('create-button'));
    expect(props.createTodo).toBeCalledWith({ title: 'title', description: 'description', status: TodoModel.Status.PENDING });
  });

  it('should reset form on focus', () => {
    fireEvent.changeText(wrapper.getByTestId('title-input'), 'title');
    fireEvent.changeText(wrapper.getByTestId('description-input'), 'description');
    navRef.current.dispatch('TestScreen2'); // leaving screen and reentering to trigger willFocus
    navRef.current.dispatch('TestScreen'); // leaving screen and reentering to trigger willFocus
    expect(wrapper.getByTestId('title-input').props.value).toEqual('');
    expect(wrapper.getByTestId('description-input').props.value).toEqual('');
  });

  it('should go to list', () => {
    fireEvent.press(wrapper.getByTestId('go-to-list-button'));
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
