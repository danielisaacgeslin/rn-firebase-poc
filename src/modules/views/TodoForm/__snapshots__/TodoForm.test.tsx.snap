// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoForm should go to list 1`] = `
<div>
  todo list
</div>
`;

exports[`TodoForm should render 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
      "paddingBottom": 20,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
          "paddingHorizontal": 20,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "space-between",
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "color": "#1745D5",
              "fontSize": 23,
              "fontWeight": "600",
              "marginBottom": 8,
              "marginTop": 16,
              "textAlign": "left",
            }
          }
        >
          Create a Todo
        </Text>
        <View
          style={
            Object {
              "marginBottom": 32,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#9FA7BC",
                "fontSize": 13,
                "fontWeight": "600",
                "marginBottom": 10,
              }
            }
          >
            title
          </Text>
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Enter your title..."
            placeholderTextColor="#D5DEF1"
            rejectResponderTermination={true}
            style={
              Object {
                "borderColor": "#D5DEF1",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#646565",
                "fontSize": 15,
                "fontWeight": "400",
                "padding": 16,
              }
            }
            testID="title-input"
            underlineColorAndroid="rgba(0, 0, 0, 0)"
            value=""
          />
        </View>
        <View
          style={
            Object {
              "marginBottom": 32,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#9FA7BC",
                "fontSize": 13,
                "fontWeight": "600",
                "marginBottom": 10,
              }
            }
          >
            description
          </Text>
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            placeholder="Enter your description..."
            placeholderTextColor="#D5DEF1"
            rejectResponderTermination={true}
            style={
              Object {
                "borderColor": "#D5DEF1",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#646565",
                "fontSize": 15,
                "fontWeight": "400",
                "padding": 16,
              }
            }
            testID="description-input"
            underlineColorAndroid="rgba(0, 0, 0, 0)"
            value=""
          />
        </View>
      </View>
      <View>
        <View
          style={
            Object {
              "borderRadius": 3,
            }
          }
        >
          <View
            accessible={true}
            clickable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="create-button"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "hsl(208, 8%, 90%)",
                  "borderColor": "#2089dc",
                  "borderRadius": 3,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": Object {
                      "color": Array [
                        208,
                        8,
                        63,
                      ],
                      "model": "hsl",
                      "valpha": 1,
                    },
                    "fontSize": 18,
                    "paddingBottom": 1,
                    "paddingTop": 2,
                    "textAlign": "center",
                  }
                }
              >
                Create Todo
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "borderRadius": 3,
            }
          }
        >
          <View
            accessible={true}
            clickable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="go-to-list-button"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#2089dc",
                  "borderRadius": 3,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginTop": 15,
                  "padding": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#2089dc",
                    "fontSize": 18,
                    "paddingBottom": 1,
                    "paddingTop": 2,
                    "textAlign": "center",
                  }
                }
              >
                Go to List
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
